uat/custom-components/client-python-paho/src/grpc_client_server/grpc_control_server.py

```diff
@@ -0,0 +1,363 @@
+        """
+        Construct GRPCControlServer
+        Parameters
+        ----------
+        client - GRPCDiscoveryClient object
+        address - local gRPC service address
+        """
+        self.__logger = GRPCControlServer.logger
+        self.__shutdown_reason = ""
+        self.__stop_server_future = asyncio.Future()
+        self.__mqtt_lib: MQTTLib = None
+
+        # TODO delete suppress warning
+        self.__client = client  # pylint: disable=unused-private-member
+
+        super().__init__()
+
+        self.__server = grpc.aio.server()
+        mqtt_client_control_pb2_grpc.add_MqttClientControlServicer_to_server(self, self.__server)
+        self.__chosen_port = self.__server.add_insecure_port(address)
+
+    def get_port(self) -> int:
+        """Returns gRPC server port"""
+        return self.__chosen_port
+
+    def get_shutdown_reason(self) -> str:
+        """Returns shutdown reason"""
+        return self.__shutdown_reason
+
+    async def start_grpc_server(self):
+        """Start grpc server"""
+        await self.__server.start()
+
+    async def wait(self, mqtt_lib: MQTTLib):
+        """
+        Wait until incoming shutdown request
+        Parameters
+        ----------
+        mqtt_lib - MQTT side of the client to handler incoming requests
+        """
+        self.__mqtt_lib = mqtt_lib
+        self.__logger.info("Server awaiting termination")
+        await self.__stop_server_future
+        await self.__server.stop(None)
+        await self.__server.wait_for_termination()
+        self.__logger.info("Server termination done")
+
+    def unblock_wait(self):
+        """Unblock wait method"""
+        try:
+            self.__stop_server_future.set_result(True)
+        except asyncio.InvalidStateError:
+            pass
+
+    def close(self):
+        """Closes the gRPC server."""
+        self.unblock_wait()
+
+    async def ShutdownAgent(self, request: ShutdownRequest, context: grpc.aio.ServicerContext) -> Empty:
+        """
+        override
+        Handler of ShutdownAgent gRPC call.
+        Parameters
+        ----------
+        request - incoming request
+        context - request context
+        Returns Empty object
+        """
+        self.__shutdown_reason = request.reason
+        self.__logger.info("shutdownAgent: reason %s", self.__shutdown_reason)
+        self.unblock_wait()
+        return Empty()
+
+    async def CreateMqttConnection(  # pylint: disable=too-many-locals,too-many-branches,too-many-statements
+        self, request: MqttConnectRequest, context: grpc.aio.ServicerContext
+    ) -> MqttConnectReply:
+        """
+        override
+        Handler of CreateMqttConnection gRPC call.
+        Parameters
+        ----------
+        request - incoming request
+        context - request context
+        Returns MqttConnectReply object
+        """
+        client_id = request.clientId
+        if (client_id is None) or (len(client_id) == 0):

```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

To check if a container or sequence (string, list, tuple) is empty, use `if not val`. Do not compare its length using `if len(val) == 0`

[Learn more](https://www.python.org/dev/peps/pep-0008/#programming-recommendations#:~:text=if%20not%20seq)

Similar issue at line numbers 140, 192, 196, and 200.